src/mesh.cpp

```diff
@@ -1597,6 +1621,108 @@ extern "C" int openmc_rectilinear_mesh_set_grid(int32_t index,
   return err;
 }
 
+//! Get the cylindrical mesh grid
+extern "C" int openmc_cylindrical_mesh_get_grid(int32_t index, double** grid_x,
+  int* nx, double** grid_y, int* ny, double** grid_z, int* nz)
+{
+  if (int err = check_mesh_type<CylindricalMesh>(index))
+    return err;
+  CylindricalMesh* m =
+    dynamic_cast<CylindricalMesh*>(model::meshes[index].get());
+
+  if (m->lower_left_.dimension() == 0) {
+    set_errmsg("Mesh parameters have not been set.");
+    return OPENMC_E_ALLOCATE;
+  }
+
+  *grid_x = m->grid_[0].data();
+  *nx = m->grid_[0].size();
+  *grid_y = m->grid_[1].data();
+  *ny = m->grid_[1].size();
+  *grid_z = m->grid_[2].data();
+  *nz = m->grid_[2].size();
+
+  return 0;
+}
+
+//! Set the cylindrical mesh parameters
+extern "C" int openmc_cylindrical_mesh_set_grid(int32_t index,
+  const double* grid_x, const int nx, const double* grid_y, const int ny,
+  const double* grid_z, const int nz)
+{
+  if (int err = check_mesh_type<CylindricalMesh>(index))
+    return err;
+  CylindricalMesh* m =
+    dynamic_cast<CylindricalMesh*>(model::meshes[index].get());
+
+  m->n_dimension_ = 3;
+  m->grid_.resize(m->n_dimension_);
+
+  for (int i = 0; i < nx; i++) {
+    m->grid_[0].push_back(grid_x[i]);
+  }
+  for (int i = 0; i < ny; i++) {
+    m->grid_[1].push_back(grid_y[i]);
+  }
+  for (int i = 0; i < nz; i++) {
+    m->grid_[2].push_back(grid_z[i]);
+  }
+
+  int err = m->set_grid();
+  return err;
+}
+
+//! Get the spherical mesh grid
+extern "C" int openmc_spherical_mesh_get_grid(int32_t index, double** grid_x,
+  int* nx, double** grid_y, int* ny, double** grid_z, int* nz)
+{
+  if (int err = check_mesh_type<SphericalMesh>(index))
+    return err;
+  SphericalMesh* m =
+    dynamic_cast<SphericalMesh*>(model::meshes[index].get());
+
+  if (m->lower_left_.dimension() == 0) {
+    set_errmsg("Mesh parameters have not been set.");
+    return OPENMC_E_ALLOCATE;
+  }
+
+  *grid_x = m->grid_[0].data();
+  *nx = m->grid_[0].size();
+  *grid_y = m->grid_[1].data();
+  *ny = m->grid_[1].size();
+  *grid_z = m->grid_[2].data();
+  *nz = m->grid_[2].size();
+
+  return 0;
+}
+
+//! Set the spherical mesh parameters
+extern "C" int openmc_spherical_mesh_set_grid(int32_t index,
+  const double* grid_x, const int nx, const double* grid_y, const int ny,
+  const double* grid_z, const int nz)
+{
+  if (int err = check_mesh_type<SphericalMesh>(index))
+    return err;
+  SphericalMesh* m =
+    dynamic_cast<SphericalMesh*>(model::meshes[index].get());
+
+  m->n_dimension_ = 3;
+  m->grid_.resize(m->n_dimension_);
+
+  for (int i = 0; i < nx; i++) {
```

Comment:
Again, should appropriately reserve space before calling push_back many times. Will be much faster!